packages/store/ts/config/v2/store.ts

```diff
@@ -22,44 +23,43 @@ export type validateStoreTablesConfig<input, scope extends AbiTypeScope = AbiTyp
 };
 
 export type resolveStoreTablesConfig<input, scope extends AbiTypeScope = AbiTypeScope> = evaluate<{
-  [key in keyof input]: resolveTableConfig<input[key], scope>;
+  readonly [key in keyof input]: resolveTableConfig<input[key], scope>;
 }>;
 
-type scopeWithUserTypes<userTypes, scope extends AbiTypeScope = AbiTypeScope> = UserTypes extends userTypes
+export type scopeWithUserTypes<userTypes, scope extends AbiTypeScope = AbiTypeScope> = UserTypes extends userTypes
   ? scope
   : userTypes extends UserTypes
     ? extendScope<scope, userTypes>
     : scope;
 
-type scopeWithEnums<enums, scope extends AbiTypeScope = AbiTypeScope> = Enums extends enums
+export type scopeWithEnums<enums, scope extends AbiTypeScope = AbiTypeScope> = Enums extends enums
   ? scope
   : enums extends Enums
     ? extendScope<scope, { [key in keyof enums]: "uint8" }>
     : scope;
 
+export type extendedScope<input> = scopeWithEnums<get<input, "enums">, scopeWithUserTypes<get<input, "userTypes">>>;
+
 export type validateStoreConfig<input> = {
   [key in keyof input]: key extends "tables"
-    ? validateStoreTablesConfig<
-        input[key],
-        scopeWithEnums<get<input, "enums">, scopeWithUserTypes<get<input, "userTypes">>>
-      >
+    ? validateStoreTablesConfig<input[key], extendedScope<input>>
     : key extends "userTypes"
       ? UserTypes
       : key extends "enums"
         ? narrow<input[key]>
         : input[key];
 };
 
+export type resolveEnums<enums> = { [key in keyof enums]: Readonly<enums[key]> };
+
 export type resolveStoreConfig<input> = evaluate<{
-  [key in keyof input]: key extends "tables"
-    ? resolveStoreTablesConfig<
-        input[key],
-        scopeWithEnums<get<input, "enums">, scopeWithUserTypes<get<input, "userTypes">>>
-      >
-    : input[key];
+  readonly tables: "tables" extends keyof input ? resolveStoreTablesConfig<input["tables"], extendedScope<input>> : {};
+  readonly userTypes: "userTypes" extends keyof input ? input["userTypes"] : {};
+  readonly enums: "enums" extends keyof input ? resolveEnums<input["enums"]> : {};
+  readonly namespace: "namespace" extends keyof input ? input["namespace"] : "";
 }>;
 
-export function resolveStoreConfig<input>(input: validateStoreConfig<input>): resolveStoreConfig<input> {
+export function resolveStoreConfig<const input>(input: validateStoreConfig<input>): resolveStoreConfig<input> {
```

Comment:
does this make it so a regular object input gets used as if it were a const/strongly typed?

Reply:
yes basically! https://xebia.com/blog/typescript-5-0-and-the-new-const-modifier-on-type-parameters/